core/coreapi/unixfile.go

```diff
@@ -23,160 +17,115 @@ const prefetchFiles = 4
 
 // TODO: this probably belongs in go-unixfs (and could probably replace a chunk of it's interface in the long run)
 
-type sizeInfo struct {
-	size    int64
-	name    string
-	modTime time.Time
-}
-
-func (s *sizeInfo) Name() string {
-	return s.name
-}
-
-func (s *sizeInfo) Size() int64 {
-	return s.size
-}
-
-func (s *sizeInfo) Mode() os.FileMode {
-	return 0444 // all read
-}
-
-func (s *sizeInfo) ModTime() time.Time {
-	return s.modTime
-}
-
-func (s *sizeInfo) IsDir() bool {
-	return false
-}
-
-func (s *sizeInfo) Sys() interface{} {
-	return nil
-}
-
 type ufsDirectory struct {
 	ctx   context.Context
 	dserv ipld.DAGService
-
-	files chan *ipld.Link
-
-	name string
-	path string
+	dir   uio.Directory
 }
 
-func (d *ufsDirectory) Close() error {
-	return files.ErrNotReader
-}
+type ufsIterator struct {
+	ctx   context.Context
+	files chan *ipld.Link
+	dserv ipld.DAGService
 
-func (d *ufsDirectory) Read(_ []byte) (int, error) {
-	return 0, files.ErrNotReader
-}
+	curName string
+	curFile files.Node
 
-func (d *ufsDirectory) FileName() string {
-	return d.name
+	err error
 }
 
-func (d *ufsDirectory) FullPath() string {
-	return d.path
+func (it *ufsIterator) Name() string {
+	return it.curName
 }
 
-func (d *ufsDirectory) IsDirectory() bool {
-	return true
+func (it *ufsIterator) Node() files.Node {
+	return it.curFile
 }
 
-func (d *ufsDirectory) NextFile() (files.File, error) {
-	l, ok := <-d.files
+func (it *ufsIterator) Next() bool {
+	l, ok := <-it.files
 	if !ok {
-		return nil, io.EOF
+		return false
 	}
 
-	nd, err := l.GetNode(d.ctx, d.dserv)
+	it.curFile = nil
+
+	nd, err := l.GetNode(it.ctx, it.dserv)
 	if err != nil {
-		return nil, err
+		it.err = err
+		return false
 	}
 
-	return newUnixfsFile(d.ctx, d.dserv, nd, l.Name, d)
+	it.curName = l.Name
+	it.curFile, it.err = newUnixfsFile(it.ctx, it.dserv, nd)
+	return it.err == nil
 }
 
-func (d *ufsDirectory) Size() (int64, error) {
-	return 0, files.ErrNotReader
+func (it *ufsIterator) Err() error {
+	return it.err
 }
 
-func (d *ufsDirectory) Seek(offset int64, whence int) (int64, error) {
-	return 0, files.ErrNotReader
+func (d *ufsDirectory) Close() error {
+	return nil
 }
 
-type ufsFile struct {
-	uio.DagReader
-
-	name string
-	path string
-}
+func (d *ufsDirectory) Entries() files.DirIterator {
+	fileCh := make(chan *ipld.Link, prefetchFiles)
+	go func() {
+		d.dir.ForEachLink(d.ctx, func(link *ipld.Link) error {
```

Comment:
What if this fails? I believe we need to stash the error.

Trivial case: context is canceled. In that case, `iterator.Err()` should return the context cancellation error.